src/modules/auto-state-on-scroll-bounds.js

```diff
@@ -0,0 +1,187 @@
+	
+	var getBounds = function (elems) {
+		var b = {
+			x: 0,
+			y: 0
+		};
+		
+		elems.each(function () {
+			var t = $(this);
+			var x = t.outerWidth() + t.offset().left;
+			var y = t.outerHeight() + t.offset().top;
+			
+			if (x > b.x) {
+				b.x = x;
+			}
+			
+			if (y > b.y) {
+				b.y = y;
+			}
+		});
+		
+		b.x = Math.floor(b.x);
+		b.y = Math.floor(b.y);
+		
+		return b;
+	};
+	
+	var getData = function (elem) {
+		var ctn = elem.closest(SELECTOR_CTN);
+		var curX = Math.floor(elem.scrollLeft());
+		var curY = Math.floor(elem.scrollTop());
+		var elemW = Math.floor(elem.outerWidth());
+		var elemH = Math.floor(elem.outerHeight());
+		var scrollW = Math.floor(elem.prop('scrollWidth'));
+		var scrollH = Math.floor(elem.prop('scrollHeight'));
+		var isScrollableX = scrollW > elemW;
+		var isScrollableY = scrollH > elemH;
+		var thresholdValue = !!elem.attr(ATTR_THRESHOLD) ?
+			parseFloat(elem.attr(ATTR_THRESHOLD)) : defaultThreshold;
+		var xThreshold = Math.floor(scrollW * thresholdValue);
+		var yThreshold = Math.floor(scrollH * thresholdValue);
+		
+		
+		var xStartVisible = !isScrollableX || curX <= xThreshold;
+		var xEndVisible = !isScrollableX || (scrollW - (curX + elemW)) <= xThreshold;
+		var yStartVisible = !isScrollableY || curY <= yThreshold;
+		var yEndVisible = !isScrollableY || (scrollH - (curY + elemH)) <= yThreshold;
+		
+		elem.data('scroll-bounds-callback', function () {
+			App.modules.notify('changeState.update', {
+				item: ctn,
+				state: 'x-start',
+				action: xStartVisible ? 'on' : 'off'
+			});
+			App.modules.notify('changeState.update', {
+				item: ctn,
+				state: 'x-end',
+				action: xEndVisible ? 'on' : 'off'
+			});
+			App.modules.notify('changeState.update', {
+				item: ctn,
+				state: 'y-start',
+				action: yStartVisible ? 'on' : 'off'
+			});
+			App.modules.notify('changeState.update', {
+				item: ctn,
+				state: 'y-end',
+				action: yEndVisible ? 'on' : 'off'
+			});
+		});
+	};
+	
+	var updateData = function (elem) {
+		App.callback(elem.data('scroll-bounds-callback'));
+	};
+	
+	// SCROLL
+	var onElementScroll = function () {
+		var t = $(this);
+		
+		getData(t);
+		
+		window.craf(scrollTimer);
+		scrollTimer = window.raf(function () {
+			updateData(t);
+		});
+	};
+	
+	var refreshElements = function () {
+		//detach event on old elements
+		elements.off('scroll', onElementScroll);
+		
+		elements = site.find(SELECTOR_SCROLL);
+		
+		//attach events on new element
+		elements.on('scroll', onElementScroll);
+	};
+
+	// RESIZE
+	var onResize = function () {
+		window.craf(resizeTimer);
+
+		resizeTimer = window.raf(function () {
+			elements.each(function () {
+				var t = $(this);
+				getData(t);
+				updateData(t);
+			});
+		});
+	};
+
+	// PAGE / ARTICLE CHANGER EVENTS
+	var refreshAll = function () {
+		refreshElements();
+		onResize();
+	};
+	
+	var init = function () {
+		setTimeout(onResize, 1000);
+		setTimeout(onResize, 2000);
```

Comment:
Why not listen for page loaded ?

Reply:
Good question. I know we sometime have problems with less not being ready in time but maybe @beaubbe has a better answer.

Reply:
YEah, I guess we could listen to page loaded

Reply:
Done, please check if it's like you wanted. @nitriques